pkg/indexer/indexer.go

```diff
@@ -195,20 +220,120 @@ The only non-retriable errors should be things like malformed events or failed v
 */
 func indexLogs(
 	ctx context.Context,
+	client blockchain.ChainClient,
 	eventChannel <-chan types.Log,
+	reorgChannel chan<- uint64,
 	logger *zap.Logger,
 	logStorer storer.LogStorer,
 	blockTracker IBlockTracker,
+	reorgHandler ChainReorgHandler,
 ) {
-	var err storer.LogStorageError
+	var (
+		errStorage         storer.LogStorageError
+		storedBlockNumber  uint64
+		storedBlockHash    []byte
+		lastBlockSeen      uint64
+		reorgCheckInterval uint64 = 10 // TODO: Adapt based on blocks per batch
+		reorgCheckAt       uint64
+		reorgDetectedAt    uint64
+		reorgBeginsAt      uint64
+		reorgFinishesAt    uint64
+		reorgInProgress    bool
+	)
+
 	// We don't need to listen for the ctx.Done() here, since the eventChannel will be closed when the parent context is canceled
 	for event := range eventChannel {
+		// 1.1 Handle active reorg state first
+		if reorgDetectedAt > 0 {
+			// Under a reorg, future events are no-op
+			if event.BlockNumber >= reorgDetectedAt {
+				logger.Debug("discarding future event due to reorg",
+					zap.Uint64("eventBlockNumber", event.BlockNumber),
+					zap.Uint64("reorgBlockNumber", reorgBeginsAt))
+				continue
+			}
+			logger.Info("starting processing reorg",
+				zap.Uint64("eventBlockNumber", event.BlockNumber),
+				zap.Uint64("reorgBlockNumber", reorgBeginsAt))
+
+			// When all future events have been discarded, it means we've reached the reorg point
+			storedBlockNumber, storedBlockHash = blockTracker.GetLatestBlock()
+			lastBlockSeen = event.BlockNumber
+			reorgDetectedAt = 0
+			reorgInProgress = true
+		}
```

Comment:
how does this work in HA? There are multiple indexer processes running. Are they somehow guaranteed to see the same reorg?